### PR TITLE
Refactor _load_checkpoint fn

### PR DESCRIPTION
**Pull request by hhaAndroid**
Refactor _load_checkpoint fn

### CODE REVIEW

File: mmcv/runner/checkpoint.py [hunk 1]
```diff
@@ -221,31 +221,141 @@ def _process_mmcls_checkpoint(checkpoint):
...
-    Returns:
-        dict | OrderedDict: The loaded checkpoint. It can be either an
-            OrderedDict storing model weights or a dict containing other
-            information, which depends on the checkpoint.
+    @classmethod
+    def load_checkpoint(cls, filename, map_location):
+        raise NotImplementedError()
+
+
+class NativeLoadCheckpointHandler(LoadCheckpointHandler):
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:240]: 1. CheckpointLoader
2. Use registry.

File: mmcv/runner/checkpoint.py [hunk 2]
```diff
@@ -221,31 +221,141 @@ def _process_mmcls_checkpoint(checkpoint):
...
+
+    @classmethod
+    def load_checkpoint(cls, filename, map_location):
+        if not osp.isfile(filename):
+            raise IOError(f'{filename} is not a checkpoint file')
+        checkpoint = torch.load(filename, map_location=map_location)
+        return checkpoint
+
+
+class LoadCheckpointHandlerClient:
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:250]: CheckpointLoaderClient

File: mmcv/runner/checkpoint.py [hunk 3]
```diff
@@ -221,31 +221,141 @@ def _process_mmcls_checkpoint(checkpoint):
     return new_checkpoint
 
 
-def _load_checkpoint(filename, map_location=None):
-    """Load checkpoint from somewhere (modelzoo, file, url).
+class LoadCheckpointHandler:
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:224]: BaseCheckpointLoader

File: mmcv/runner/checkpoint.py [hunk 4]
```diff
@@ -221,31 +221,141 @@ def _process_mmcls_checkpoint(checkpoint):
...
+                return cls._handler[p]
+        return cls._native_checkpoint_handler
+
+    @classmethod
+    def load_checkpoint(cls, filepath, map_location=None):
+        return cls._get_checkpoint_handler(filepath).load_checkpoint(
+            filepath, map_location)
+
+
+@LoadCheckpointHandlerClient.register_handler()
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:317]: Can we use, thus we do not rely on `get_supported_prefixes`.
```python
@LoadCheckpointHandlerClient.register_handler(prefixes['http', 'https'])
```

File: mmcv/runner/checkpoint.py [hunk 5]
```diff
@@ -302,7 +438,8 @@ def load_checkpoint(model,
     Returns:
         dict or OrderedDict: The loaded checkpoint.
     """
-    checkpoint = _load_checkpoint(filename, map_location)
+    checkpoint = LoadCheckpointHandlerClient.load_checkpoint(
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:441]: CheckpointLoaderClient.load_checkpoint()

File: mmcv/runner/checkpoint.py [hunk 6]
```diff
@@ -221,31 +221,141 @@ def _process_mmcls_checkpoint(checkpoint):
...
+            else:
+                raise KeyError(
+                    f'{handler} is already registered as a storage backend, '
+                    'add "force=True" if you want to override it')
+        # sort
+        cls._handler = OrderedDict(
+            sorted(cls._handler.items(), key=lambda t: t[0], reverse=True))
+
+    @classmethod
+    def register_handler(cls, handler=None, force=False):
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:283]: I do not get the logic here.

**Comment by ZwwWayne** [comment]
CI failed. May skip the unit test of loading when the torch package is unaccessible.

File: mmcv/runner/checkpoint.py [hunk 7]
```diff
@@ -221,31 +221,120 @@ def _process_mmcls_checkpoint(checkpoint):
...
+    @classmethod
+    def _register_loader(cls, prefixes, loader, force=False):
+        assert isinstance(prefixes, (list, tuple, str))
+        if isinstance(prefixes, str):
+            prefixes = [prefixes]
+        for prefix in prefixes:
+            if prefix not in cls._loader:
+                cls._loader[prefix] = loader
+                continue
+            if force:
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:262]: may use `elif`?

File: mmcv/runner/checkpoint.py [hunk 8]
```diff
@@ -221,31 +221,120 @@ def _process_mmcls_checkpoint(checkpoint):
...
+        """Finds a loader that supports the given path. Falls back to the
+        native loader if no other loader is found.
+
+        Args:
+            path (str): checkpoint path
+
+        Returns:
+            loader (Loader)
+        """
+        for p in cls._loader.keys():
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:296]: cls._loader.keys() -> cls._loader

File: mmcv/runner/checkpoint.py [hunk 9]
```diff
@@ -221,31 +221,120 @@ def _process_mmcls_checkpoint(checkpoint):
...
+            if path.startswith(p):
+                return cls._loader[p]
+        return cls._native_checkpoint_loader
+
+    @classmethod
+    def load_checkpoint(cls, filepath, map_location=None):
+        return cls._get_checkpoint_loader(filepath).load_checkpoint(
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:303]: Shall we indicate which loader is used to load the checkpoint during loading?

Comment (hhaAndroid) [reply mmcv/runner/checkpoint.py:303]: It May was not good because print information every time you call

File: tests/test_runner/test_checkpoint.py [hunk 10]
```diff
@@ -178,3 +178,86 @@ def test_load_classes_name():
...
+    loader = CheckpointLoaderClient._get_checkpoint_loader(filename)
+    assert loader.__name__ == 'FTPCheckpointLoader'
+
+    # test force
+    class FTP1CheckpointLoader(BaseCheckpointLoader):
+
+        @classmethod
+        def load_checkpoint(cls, filename, map_location):
+            return dict(filename=filename)
+
```
Comment (ZwwWayne) [inline tests/test_runner/test_checkpoint.py:250]: Add a comment on which error you are testing.

File: mmcv/runner/checkpoint.py [hunk 11]
```diff
@@ -221,31 +221,204 @@ def _process_mmcls_checkpoint(checkpoint):
...
+            raise IOError(f'{filename} is not a checkpoint file')
+        checkpoint = torch.load(filename, map_location=map_location)
+        return checkpoint
+
+
+class CheckpointLoaderClient:
+    """A general checkpoint loader client to load checkpoint.
+
+    Attributes:
+        _loader (dict[loader]): .
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:261]: ?

File: mmcv/runner/checkpoint.py [hunk 12]
```diff
@@ -221,31 +221,204 @@ def _process_mmcls_checkpoint(checkpoint):
...
+    @classmethod
+    def _get_checkpoint_loader(cls, path):
+        """Finds a loader that supports the given path. Falls back to the
+        native loader if no other loader is found.
+
+        Args:
+            path (str): checkpoint path
+
+        Returns:
+            loader (Loader)
```
Comment (ZwwWayne) [inline mmcv/runner/checkpoint.py:322]: There is no such a type named `Loader`, may use 
```python
cls:`BaseCheckpointLoader`
```

**Comment by hellock** [comment]
The encapsulation is a little complicated and could be simpler:

```python
def load_checkpoint():
    # use CheckpointLoader to load from anywhere
    pass


# manage all scheme loaders
class CheckpointLoader:

    @classmethod
    def register_scheme(cls):
        pass

    @classmethod
    def load_checkpoint(cls):
        pass


CheckpointLoader.register_scheme(prefixes=('http://', 'https://'))
def load_from_http():
    pass

CheckpointLoader.register_scheme(prefixes=('modelzoo://', 'torchvision://'))
def load_from_torchvision():
    pass

```

File: mmcv/runner/checkpoint.py [hunk 13]
```diff
@@ -221,67 +221,233 @@ def _process_mmcls_checkpoint(checkpoint):
...
+        else:
+            assert isinstance(prefixes, (list, tuple))
+        for prefix in prefixes:
+            if (prefix not in cls._loader) or force:
+                cls._loader[prefix] = loader
+            else:
+                raise KeyError(
+                    f'{prefix} is already registered as a loader backend, '
+                    'add "force=True" if you want to override it')
+        # sort
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:259]: The comment should be why instead of what.

File: mmcv/runner/checkpoint.py [hunk 14]
```diff
@@ -221,67 +221,233 @@ def _process_mmcls_checkpoint(checkpoint):
     return new_checkpoint
 
 
-def _load_checkpoint(filename, map_location=None):
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:224]: Keep the old api to maintain backward compatibility.

File: mmcv/runner/checkpoint.py [hunk 15]
```diff
@@ -221,67 +221,233 @@ def _process_mmcls_checkpoint(checkpoint):
...
+    if not osp.isfile(filename):
+        raise IOError(f'{filename} is not a checkpoint file')
+    checkpoint = torch.load(filename, map_location=map_location)
+    return checkpoint
+
+
+class CheckpointLoader:
+    """A general checkpoint loader to manage all scheme loaders."""
+
+    _loader = {}
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:244]: _schemes

File: mmcv/runner/checkpoint.py [hunk 16]
```diff
@@ -221,67 +221,233 @@ def _process_mmcls_checkpoint(checkpoint):
...
+        dict or OrderedDict: The loaded checkpoint.
+    """
+
+    model_path = filename[7:]
+    checkpoint = load_pavimodel_dist(model_path, map_location=map_location)
+    return checkpoint
+
+
+@CheckpointLoader.register_scheme(prefixes='s3://')
+def load_from_s3(filename, map_location):
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:435]: `load_from_ceph`

File: tests/test_runner/test_checkpoint.py [hunk 17]
```diff
@@ -178,3 +178,78 @@ def test_load_classes_name():
...
+    loader = CheckpointLoader._get_checkpoint_loader(filename)
+    assert loader.__name__ == 'load_from_pavi'
+
+    filename = 's3://xx.xx/xx.pth'
+    loader = CheckpointLoader._get_checkpoint_loader(filename)
+    assert loader.__name__ == 'load_from_s3'
+
+    filename = 'ss3://xx.xx/xx.pth'
+    loader = CheckpointLoader._get_checkpoint_loader(filename)
+    assert loader.__name__ == 'load_from_local'
```
Comment (hellock) [inline tests/test_runner/test_checkpoint.py:230]: Use for-loop test the above cases.

File: mmcv/runner/checkpoint.py [hunk 18]
```diff
@@ -221,67 +221,253 @@ def _process_mmcls_checkpoint(checkpoint):
...
+    Args:
+        filename (str): checkpoint file path with pavi prefix
+        map_location (str, optional): Same as :func:`torch.load`.
+
+    Returns:
+        dict or OrderedDict: The loaded checkpoint.
+    """
+
+    model_path = filename[7:]
+    checkpoint = load_pavimodel_dist(model_path, map_location=map_location)
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:430]: `load_from_pavi` and `load_pavimodel_dist` can be merged.

File: mmcv/runner/checkpoint.py [hunk 19]
```diff
@@ -221,67 +221,253 @@ def _process_mmcls_checkpoint(checkpoint):
...
+
+    Args:
+        filename (str): checkpoint file path with s3 prefix
+        map_location (str, optional): Same as :func:`torch.load`.
+
+    Returns:
+        dict or OrderedDict: The loaded checkpoint.
+    """
+
+    checkpoint = load_fileclient_dist(
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:446]: `load_from_ceph` and `load_fileclient_dist` can be merged.

File: mmcv/runner/__init__.py [hunk 20]
```diff
@@ -25,13 +25,13 @@
...
-    'RUNNERS', 'allreduce_grads', 'allreduce_params', 'LossScaler'
+    'WandbLoggerHook', 'MlflowLoggerHook', 'load_state_dict',
+    'load_checkpoint', 'weights_to_cpu', 'save_checkpoint', 'Priority',
+    'get_priority', 'get_host_info', 'get_time_str', 'obj_from_dict',
+    'init_dist', 'get_dist_info', 'master_only', 'OPTIMIZER_BUILDERS',
+    'OPTIMIZERS', 'DefaultOptimizerConstructor', 'build_optimizer',
+    'build_optimizer_constructor', 'IterLoader', 'set_random_seed',
+    'auto_fp16', 'force_fp32', 'wrap_fp16_model', 'Fp16OptimizerHook',
+    'SyncBuffersHook', 'EMAHook', 'build_runner', 'RUNNERS', 'allreduce_grads',
+    'allreduce_params', 'LossScaler', 'CheckpointLoader', '_load_checkpoint'
```
Comment (hellock) [inline mmcv/runner/__init__.py:36]: The order can be kept to avoid unnecessary changes.

File: mmcv/runner/checkpoint.py [hunk 21]
```diff
@@ -105,23 +105,206 @@ def load(module, prefix=''):
...
+    new_state_dict = OrderedDict()
+    for k, v in state_dict.items():
+        if k.startswith('backbone.'):
+            new_state_dict[k[9:]] = v
+    new_checkpoint = dict(state_dict=new_state_dict)
+
+    return new_checkpoint
+
+
+def load_from_local(filename, map_location):
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:161]: We can register a scheme of empty string for `load_from_local` to unify all cases.

```python
@CheckpointLoader.register_scheme(prefixes='')
def load_from_local(filename, map_location):
    pass
```

File: mmcv/runner/checkpoint.py [hunk 22]
```diff
@@ -168,118 +362,98 @@ def load_fileclient_dist(filename, backend, map_location):
...
+        warnings.warn('The URL scheme of "modelzoo://" is deprecated, please '
+                      'use "torchvision://" instead')
+        model_name = filename[11:]
+    else:
+        model_name = filename[14:]
+    return load_from_http(model_urls[model_name])
 
-    return default_urls
 
+@CheckpointLoader.register_scheme(prefixes='open-mmlab://')
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:388]: We can register both `openmmlab://` and `open-mmlab://`.

File: mmcv/runner/__init__.py (hunk 20 above)
Comment (hhaAndroid) [reply mmcv/runner/__init__.py:36]: This is the correct order before modification and can be kept

File: mmcv/runner/checkpoint.py [hunk 23]
```diff
@@ -168,118 +362,100 @@ def load_fileclient_dist(filename, backend, map_location):
...
+        map_location (str, optional): Same as :func:`torch.load`.
+          Default: None
 
-    return mmcls_urls
+    Returns:
+        dict or OrderedDict: The loaded checkpoint.
+    """
 
+    model_urls = get_external_models()
+    model_name = filename[13:]
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:404]: `open-mmlab://` and `openmmlab://` are different.

File: mmcv/runner/__init__.py (hunk 20 above)
Comment (hellock) [reply mmcv/runner/__init__.py:36]: Seems that `_load_checkpoint` are moved to the last.

Comment (hhaAndroid) [reply mmcv/runner/__init__.py:36]: Ok

**Review by hellock — APPROVED** [review]
